### PR TITLE
Add system reference values to manifest parsing

### PR DESCRIPTION
**Pull request by JohnMcPMS**
## Change
Adds the PackageFamilyName and ProductCode values to parsing by the manifest.  These can be at the root level, where they will be inherited.  They can also be provided per installer, but they must match the InstallerType or it will be an error.

Also makes major changes to the way the ManifestVersion is handled.  For now, the version is largely ignored (except it must be major version 0 if provided).  A concept of extensions is added, which are defined as the hyphen separated values after the primary version.  These can also be versioned, although the code does not exist to check it.  The syntax is:
```
ManifestVersion := Version (- Extension)
Extension := Version (- Extension)
```
where the Version portion of Extension is treated as `Name.Version`.  For instance, `1.0.0-extension` and `1.0.0-extension.2`.

Additionally:
- Fixes an issue with the version sorting where additional version characters would sort higher than those without (ex. 1.0 will now be greater than 1.0-alpha)

## Validation
Tests are added for PackageFamilyName and ProductCode, as well as the manifest version parsing.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/winget-cli/pull/574)

### CODE REVIEW

File: src/AppInstallerCLITests/AppInstallerCLITests.vcxproj [hunk 1]
```diff
@@ -409,6 +409,18 @@
     <CopyFileToFolders Include="TestData\Manifest-Encoding-UTF8.yaml">
       <DeploymentContent>true</DeploymentContent>
     </CopyFileToFolders>
+    <CopyFileToFolders Include="TestData\Manifest-Bad-PackageFamilyNameOnMSI.yaml">
+      <DeploymentContent Condition="'$(Configuration)|$(Platform)'=='Debug|x64'">true</DeploymentContent>
```
Comment (yao-msft) [inline src/AppInstallerCLITests/AppInstallerCLITests.vcxproj:413]: ><DeploymentContent Condition="'$(Configuration)|$(Platform)'=='Debug|x64'">true</DeploymentContent> [](start = 6, length = 99)

These should apply to all. And better keep them consistent with all above

File: src/AppInstallerCLITests/YamlManifest.cpp [hunk 2]
```diff
@@ -226,7 +226,9 @@ TEST_CASE("ReadBadManifests", "[ManifestValidation]")
         { "Manifest-Bad-VersionInvalid.yaml", "Invalid field value. Field: Version" },
         { "Manifest-Bad-VersionMissing.yaml", "Required field missing. Field: Version" },
         { "Manifest-Bad-InvalidManifestVersionValue.yaml", "Invalid field value. Field: ManifestVersion" },
-        { "InstallFlowTest_MSStore.yaml", "Field value is not supported. Field: InstallerType Value: MSStore" },
```
Comment (yao-msft) [inline src/AppInstallerCLITests/YamlManifest.cpp:229]: >Field value is not supported. Field: InstallerType Value: MSStore [](start = 43, length = 65)

Is it intentional to remove this error message? I think it's good and informational to manifest authors

**Comment by yao-msft** [comment]
> TEST_CASE("ReadGoodManifestAndVerifyContents", "[ManifestValidation]")

This test was supposed to a complete content verification on a complete manifest. Since you updated Manifest_Good.yaml, maybe check the 2 new values here too

---
Refers to: src/AppInstallerCLITests/YamlManifest.cpp:31 in 97ba58f. [](commit_id = 97ba58f41035aa056a65802af9cc1f792d985c1e, deletion_comment = False)

File: src/AppInstallerCommonCore/Manifest/Manifest.cpp [hunk 3]
```diff
@@ -6,20 +6,53 @@
...
+                if (!m_parts[i].Other.empty())
+                {
+                    validationSuccess = false;
+                    break;
+                }
+            }
+
+            for (const Version& ext : m_extensions)
+            {
+                if (ext.GetParts().empty() || ext.GetParts()[0].Integer != 0)
```
Comment (yao-msft) [inline src/AppInstallerCommonCore/Manifest/Manifest.cpp:55]: >[0].Integer != 0 [](start = 60, length = 16)

Are we going to do anything about the remaining parts in extensions?

File: src/AppInstallerCommonCore/Manifest/YamlParser.cpp [hunk 4]
```diff
@@ -260,6 +253,28 @@ namespace AppInstaller::Manifest
...
+            // Validate system reference strings if they are set at the installer level
+            if (fullValidation && !installer.PackageFamilyName.empty() && !ManifestInstaller::DoesInstallerTypeUsePackageFamilyName(installer.InstallerType))
+            {
+                resultErrors.emplace_back(ManifestError::InstallerTypeDoesNotSupportPackageFamilyName, "InstallerType", ManifestInstaller::InstallerTypeToString(installer.InstallerType));
+            }
+
+            if (fullValidation && !installer.ProductCode.empty() && !ManifestInstaller::DoesInstallerTypeUseProductCode(installer.InstallerType))
+            {
+                resultErrors.emplace_back(ManifestError::InstallerTypeDoesNotSupportProductCode, "InstallerType", ManifestInstaller::InstallerTypeToString(installer.InstallerType));
+            }
```
Comment (yao-msft) [inline src/AppInstallerCommonCore/Manifest/YamlParser.cpp:265]: In my previous change to separate manifest population vs manifest semantic validation, these checks fit in the ValidateManifest() in manifestvalidation.cpp

**Comment by yao-msft** [comment]
Since we are not checking minor versions any more, do you expect us to do strict minor version after 1.0? Or minor version is no longer used as all?

**Comment by JohnMcPMS** [comment]
I don't see the value in being strict about minor version.  It seems to me to only serve to increase complexity for users adding manifests without adding value.  If I'm missing something, I am happy to become more learned, but I currently don't see any point now or in the future.

---
In reply to: [694644474](https://github.com/microsoft/winget-cli/pull/574#issuecomment-694644474) [](ancestors = 694644474)

File: src/AppInstallerCLITests/YamlManifest.cpp (hunk 2 above)
Comment (JohnMcPMS) [reply src/AppInstallerCLITests/YamlManifest.cpp:229]: I believe the problem comes from the failure in ManifestVersion completely blocking all other validation.  I don't think we can proceed if we don't like the version.  But I can drop the fullValidation from ManifestVer and allow it to get back to the original error here.

---
In reply to: [490678974](https://github.com/microsoft/winget-cli/pull/574#discussion_r490678974) [](ancestors = 490678974)

File: src/AppInstallerCommonCore/Manifest/Manifest.cpp (hunk 3 above)
Comment (JohnMcPMS) [reply src/AppInstallerCommonCore/Manifest/Manifest.cpp:55]: What do you mean "do anything"?  Validate?  I don't see any value to doing so.  Again, happy to be shown a reason and change it.

---
In reply to: [490692976](https://github.com/microsoft/winget-cli/pull/574#discussion_r490692976) [](ancestors = 490692976)

File: src/AppInstallerCommonCore/Manifest/YamlParser.cpp (hunk 4 above)
Comment (JohnMcPMS) [reply src/AppInstallerCommonCore/Manifest/YamlParser.cpp:265]: Agreed, *but* this check can only be performed in this way here.  But I can go the stricter route and move them.

---
In reply to: [490695526](https://github.com/microsoft/winget-cli/pull/574#discussion_r490695526) [](ancestors = 490695526)

Comment (JohnMcPMS) [reply src/AppInstallerCommonCore/Manifest/YamlParser.cpp:265]: Actually, I was wrong, and simply moving the validation parts will have the same net effect.

---
In reply to: [491164478](https://github.com/microsoft/winget-cli/pull/574#discussion_r491164478) [](ancestors = 491164478,490695526)

**Review by yao-msft — APPROVED** [review]
:shipit:

**Pull request merged**